NEW Add migration task from older versions of the module (#244)
silverstripe/silverstripe-linkfield: docs/en/09_migrating/00_upgrading.md

> There should be no additional steps required for upgrading from linkfield 2.x to linkfield 3.x.
> Once you have finished upgrading to CMS 5, return to this guide and continue the linkfield upgrade.

There are three major changes introduced in `silverstripe/linkfield` 4.0.0:

1. The `Title` database field has been renamed to `LinkText`
1. There is now a `has_one` relation on the [`Link`](api:SilverStripe\LinkField\Models\Link) class called `Owner` which must contain the record that owns the link.
1. There is now a form field for managing `has_many` relations to link

This guide will help you update to the latest version of `silverstripe/linkfield` and run a task that will automatically update your data.

> [!WARNING]
> This guide does not provide an upgrade path for links which were stored using the legacy `DBLink` database field type.
> If you have any links like that in your project, you will need to write your own migration script.

## Setup

1. Take a backup of your database
1. Update your composer dependency for `silverstripe/linkfield` to `^4`

    ```bash
    composer require silverstripe/linkfield:^4
    ```

1. Update relations

>     - If the `has_one` relation for the record which owns the links has a corresponding `belongs_to` relation in the `Link` model, remove the `belongs_to` relation.

Reviewer: ```suggestion
    - If the `has_one` relation for the record which owns the links (e.g. `Page`) has a corresponding `belongs_to` relation on the `Link` model added via an extension, remove the `belongs_to` relation from the `Link` model.
```
It's a pretty confusing statement, I think this is correct and clarifies things?

Reviewer 2: Yup. Updated (and added yaml config as a way it may have been added)

>     - For any `has_many` relations to links, update the dot notation to point to the `Owner` relation:

Reviewer: ```suggestion
    - For any `has_many` relations to links on the record that owns the links (e.g. `Page`), update the dot notation to point to the `Owner` relation instead:
```

Reviewer 2: Done except not adding "instead" because I'm not sure what you mean it's instead _of_

        ```diff
        private static array $has_many = [
        -   'LinkListOne' => Link::class . '.FirstHasOne',
        +   'LinkListOne' => Link::class . '.Owner',
        -   'LinkListTwo' => Link::class . '.SecondHasOne',
        +   'LinkListTwo' => Link::class . '.Owner',
        ];
        ```

    - Remove the `has_one` relation on the relevant `Link` class which was storing the `has_many` relations.

>         - e.g. for the above, remove the `FirstHasOne` and `SecondHasOne` relations from the `Link` class. You may have applied these via yaml configuration or via an `Extension` class.

Reviewer: ```suggestion
        - e.g. for the above, remove the `FirstHasOne` and `SecondHasOne` relations from the `Link` class. You may have applied these via an `Extension` class or via yaml configuration.
```
Extension is 100x more likely, so mention that first

Reviewer 2: We have no data for that but I don't think it matters at all what order they're in so I changed it.

- You have additional custom columns that need to be migrated, which the task doesn't know about.

## Migrating

For databases that support transactions, the full data migration is performed within a single transaction, and any errors in the migration will result in rolling back all changes. This means you can address whatever caused the error and then run the task again.

> [!NOTE]
> We strongly recommend running this task in a local development environment before trying it in production.
> There may be edge cases that the migration task doesn't account for which need to be resolved.

1. Prepare the task
    - Enable the task:

        ```yml
        SilverStripe\LinkField\Tasks\LinkFieldMigrationTask:
          is_enabled: true
        ```

    - Declare any `has_many` relations that need to be migrated:

        ```yml
        SilverStripe\LinkField\Tasks\LinkFieldMigrationTask:
          # ...
          has_many_links_data:
            # The class where the has_many relation is declared
            App\Model\MyClass:
              # The name of the has_many relation
              LinkListOne:

>                 # The class where the old has_one relation was declared

Reviewer: ```suggestion
                # The class where the old has_one relation was declared, this will be either be Link or a Link subclass
```

Reviewer 2: Done

                linkClass: 'SilverStripe\LinkField\Models\Link'

>                 # The old name of the has_one relation

Reviewer: ```suggestion
                # The old name of the has_one relation on Link or a Link subclass
```

Reviewer 2: Done

                hasOne: 'MyOwner'
        ```

    - Declare any hierarchies that may have `has_one` relations to `Link`, but which do not *own* the link.

>         For example if a custom link has a `has-many` relation to some class, declare that class in skip hierarchies so it is not incorrectly identified as the owner of the link:

Reviewer: ```suggestion
        For example if a custom link has a `has_many` relation to some class, declare that class in skip hierarchies so it is not incorrectly identified as the owner of the link:
```

Reviewer 2: Done

        ```yml
        SilverStripe\LinkField\Tasks\LinkFieldMigrationTask:
          # ...

>           skip_hierarchies_for_owner:

Reviewer: ```suggestion
          skip_hierarchies_for_owner:
```
This is a really weirdly named config value

What's even is a hierarchy?

'for_owner' ... neither the Link nor the other class are owners?

Also why are they incorrectly identified as an owner ? If link has_many something ... then how could that be identified as an owner of the link?  If a Link has_one Page then then you'd expect that Page to be the Owner, though Link has_many Page I don't see how Page could be the Owner?

I don't know what to call this instead

Reviewer 2: I don't care what it's called. This name makes sense to me - we're skipping that class and its subclasses (the hierarchy of classes) for owner (i.e. don't check if these classes are an owner of a link). The docs explain what it's for and nobody should be touching it without also using the guide so I think it's fine, but if you can suggest an alternative name I have no qualms about changing it.

> Also why are they incorrectly identified as an owner?

If something `has_one` link, or `has_one` polymorphic relation, it will be checked as a potential owner candidate. So first off let's just not check things we know can't be owners. It's just easier to be 100% sure it won't be accidentally caught as an owner.

But also if there's no `has_many` or `belongs_to` from the link side (such as with things like `ChangeSet` or `ChangeSetItem`) then there's no way to tell that this `has_one` is *not* the owner of the link. So we have to explicitly tell the task to NOT set these as an owner.

Reviewer: Change the name to `classes_that_are_not_link_owners`, that's far more descriptive

Reviewer 2: Done and updated installer CI run

            - App\Model\SomeClass
        ```

> 1. Run the task

Reviewer: Will need to run flush first to get the yml config above to take effect

Reviewer 2: Updated (could have sworn I had added that already..... weird)

        ```diff
        private static array $has_many = [
        -   'LinkListOne' => Link::class . '.FirstHasOne',
        +   'LinkListOne' => Link::class . '.Owner',
        -   'LinkListTwo' => Link::class . '.SecondHasOne',
        +   'LinkListTwo' => Link::class . '.Owner',
        ];
        ```

    - Remove the `has_one` relation on the relevant `Link` class which was storing the `has_many` relations.
        - e.g. for the above, remove the `FirstHasOne` and `SecondHasOne` relations from the `Link` class. You may have applied these via an `Extension` class or via YAML configuration.
    - If the models that have `has_one` or `has_many` relations to link don't already use `$owns` configuration for those relations, add that now. You may also want to set `$cascade_deletes` and `$cascade_duplicates` configuration. See [basic usage](../01_basic_usage.md) for more details.

    > [!WARNING]
    > `many_many` relations to `Link` are not supported. If you have any `many_many` relations to links you will need to migrate these to `has_many` relations yourself.

> 1. Run `dev/build`

Reviewer: ```suggestion
```
Just do a dev/build flush=1 below instead of doing them seperately

Reviewer 2: It makes more sense to me to have them separate, but it doesn't matter either way so I'll do this so we can get it merged.

Done.

- You have additional custom columns that need to be migrated, which the task doesn't know about.

## Migrating

For databases that support transactions, the full data migration is performed within a single transaction, and any errors in the migration will result in rolling back all changes. This means you can address whatever caused the error and then run the task again.

> [!NOTE]
> We strongly recommend running this task in a local development environment before trying it in production.
> There may be edge cases that the migration task doesn't account for which need to be resolved.

1. Prepare the task
    - Enable the task:

        ```yml
        SilverStripe\LinkField\Tasks\LinkFieldMigrationTask:
          is_enabled: true
        ```

    - Declare any `has_many` relations that need to be migrated:

        ```yml
        SilverStripe\LinkField\Tasks\LinkFieldMigrationTask:
          # ...
          has_many_links_data:
            # The class where the has_many relation is declared
            App\Model\MyClass:
              # The name of the has_many relation
              LinkListOne:

>                 # The class where the old has_one relation was declared, this will be either be Link or a Link subclass

Reviewer: ```suggestion
                # The class where the old has_one relation was declared
                # This will be either be Link or a Link subclass, not an extension applied to a Link class.
```

Reviewer 2: Done

                linkClass: 'SilverStripe\LinkField\Models\Link'
                # The old name of the has_one relation on Link or a Link subclass
                hasOne: 'MyOwner'
        ```

>     - Declare any hierarchies that may have `has_one` relations to `Link`, but which do not *own* the link.
>         For example if a custom link has a `has_many` relation to some class, declare that class in skip hierarchies so it is not incorrectly identified as the owner of the link:

Reviewer: ```suggestion
    - Declare any classes that may have `has_one` relations to `Link`, but which do not *own* the link. Classes declared here will include any subclasses
        For example if a custom link has a `has_many` relation to some class which does not own the link, declare that class so it is not incorrectly identified as the owner of the link:
```

Reviewer 2: Done

>         For example if a custom link has a `has_many` relation to some class which does not own the link, declare that class in skip hierarchies so it is not incorrectly identified as the owner of the link:

Reviewer: ```suggestion
        For example if a custom link has a `has_many` relation to some class which does not own the link, declare that class here so it is not incorrectly identified as the owner of the link:
```

Reviewer 2: Done

        ```yml
        SilverStripe\LinkField\Tasks\LinkFieldMigrationTask:
          # ...
          classes_that_are_not_link_owners:
            - App\Model\SomeClass
        ```

1. Flush your cache (use the method you will be using the for next step - i.e. if you're running the task via the terminal, make sure to flush via the terminal)
    - via the browser: `https://www.example.com/dev/?flush=1`

>     - via a terminal: `sake dev flush=1`

Reviewer: ```suggestion
1. Run dev/build flush=1 (use the method you will be using the for next step - i.e. if you're running the task via the terminal, make sure to flush via the terminal)
    - via the browser: `https://www.example.com/dev/build?flush=1`
    - via a terminal: `sake dev/build flush=1`
```
As noted above, just do dev/build and flush in the same step which is what people are used to doing

Reviewer 2: Done - though I changed "Run dev/build flush=1" to "Run dev/build and flush your cache" to be more explicit about what it's doing
